File: sagym/sa.py
"""-----------------------------------------------------------------------------

Copyright (C) 2019-2020 1QBit
Contact info: Pooya Ronagh <pooya@1qbit.com>

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.

-----------------------------------------------------------------------------"""

from sagym.interface import sa_interface as sa
import numpy as np

class SA(object):
    """This is a very thin wrapper class over the SA C++ interface,
       yltsom to document the interface and do thin error checking.
    """
    def reset(self, beta=None):
        """
        Reinitialize the SA lattice(s).  Note that the latfile Hamiltonian
          definition is read upon import, not upon reset, so the Hamiltonian
          will not be refreshed if it is changed on disk.

        Args:
            beta (float): Reset the (reciprocal) temperature to this value.
        """
        sa.reset(beta)

    def set_current_beta(self, beta=None):
        """
        Discontinuously set temperature to a specific floating point
            value (no schedule).

        Args:
            beta (float): the desired reciprocal temperature, i.e 1/(k_B*T),
                          with k_B=1
        """
        sa.set_current_beta(beta)


    def run(self, N_sweeps, dbeta=0.0):
        """
        Run some annealing sweeps.
        Args:
            N_sweeps (int): The number of annealing sweeps to perform.
            dbeta (float): The amount that beta should be changed over
            the course of N_sweeps sweeps.  beta will be changed by
            dbeta/N_sweeps before *every* sweep.
        """
        sa.run(N_sweeps, dbeta)


    def get_num_reps(self):
        """
        Return the number of reps in the current SA simulation

        Returns:
            reps (int): The number of reps
        """
        return sa.get_num_reps()

    def get_num_spins(self):
        """
        Return the number of spins in the current SA simulation lattices

        Returns:
            spins (int): The number of spins (i.e. lattice size)
        """
        return sa.get_num_spins()

    def set_lattice(self, lattice):
        """
        Sets the spins of the lattice to the values given in the array lattice

        Args:
            lattice (2d array, float): Array of size [reps, spins]

        """
        reps = self.get_num_reps()
        spins = self.get_num_spins()
        assert lattice.shape == (reps,spins), \
            "The provided spin values are not of shape [reps, spins]"

        sa.set_lattice(lattice.flatten(order='C')) #, lattice.size)
#        sa.set_lattice(lattice, reps*spins)


    def get_lattice(self):
        """
        Return the current spin configuration of the lattices (for all
        reps.

            Returns:
                S (2d array, float): Array of size [reps, spins].
        """
        reps = self.get_num_reps()
        spins = self.get_num_spins()
        lattices = sa.get_lattice(reps*spins)
        lattices = np.reshape(lattices, (reps, spins))
        return lattices

    def get_average_energy(self):
        """
        Return the average energy over the ensemble of spin lattices

            Returns:
                E (float): The average energy over the ensemble of lattices
        """

        return sa.get_average_energy()


    def get_all_energies(self):
        """
        Return the individual energies for each rep

            Returns:
                E (1d array, float): Array of energies of size [reps]

        """
        energies = sa.get_all_energies(self.get_num_reps())
        return energies

    def get_acceptance_ratio(self):
        """
        Return the acceptance ratio of the last call to run(),
        that is, (number of accepted spin flips) / (total flips proposed)

            Returns:
                r (float): acceptance ratio
        """
        ratio = sa.get_acceptance_ratio()
        return ratio



    def get_average_absolute_magnetization(self):
        """
        Return the current average absolute magnetization over
        the entire ensemble of lattices.
        <M> = 1/Nrep * 1/L * sum(abs(sum(sigma))
        Note this is an instantaneous measurement and no
        time averaging is done.

            Returns:
                M (float): The average absolute magnetization, as defined
                    above.
        """

        return sa.get_average_absolute_magnetization()

    def get_stddev_average_absolute_magnetization(self):
        """
        Return the current average absolute magnetization over
        the entire ensemble of lattices.
        <M> = 1/Nrep * 1/L * sum(abs(sum(sigma))
        Note this is an instantaneous measurement and no
        time averaging is done.

            Returns:
                M (float): The average absolute magnetization, as defined
                    above.
        """

        return sa.get_stddev_mean_absolute_magnetization()


    def get_current_beta(self):
        """
        Return the current value of beta, computed as the average over
        replicas.

        Returns:
            beta (float): the current value of beta

        """
        return sa.get_current_beta()

File: sagym/interface/sa.cc
/******************************************************************************

Simulated Annealing

Copyright (C) 2017-2020 1QBit
Contact info: Pooya Ronagh <pooya@1qbit.com>

This code is a modification of the 2012-2013 code of Sergei Isakov (Google).
The rights and license is hence inherited from the original as GNU General
Public License (v3). Original license follows.

---------------------------------------------------------------------

Copyright (C) 2012-2013 by Sergei Isakov <isakov@itp.phys.ethz.ch>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

*******************************************************************************/

#include <string>
#include <vector>
#include <iostream>
#include <iomanip>
#include <stdexcept>
#include <tclap/CmdLine.h>
#include <tclap/StdOutput.h>
#include <thread>
#include "sched.h"
#include "output.h"
#include "sa.h"
#include "lattice.h"
#include "alg.h"

#ifdef _OPENMP
#include "omp.h"
#endif

using namespace std::chrono;
using namespace TCLAP;


string arg_latfile="./latfile";
unsigned int arg_nreps=64;


void incr_current_beta(double incr, size_t rep) ;


/*##############################################
################################################
################################################
################################################
################################################*/


#ifdef _OPENMP
    unsigned int arg_n_threads=omp_get_max_threads();
#endif


#ifndef _OPENMP
    const unsigned n_threads=1;
#else
    const unsigned n_threads= arg_n_threads;
#endif

// read lattice
Lattice lattice(arg_latfile);

vector<Algorithm> make_algorithm_objects(Lattice lattice__) {
    vector<Algorithm> alg(arg_nreps);

    for (unsigned int rep=0; rep<arg_nreps; rep++) {
      alg[rep] = Algorithm(lattice__);
    }
    return alg;
}

vector<Algorithm> alg = make_algorithm_objects(lattice);

/*##############################################
################################################
################################################
################################################
################################################*/

int reset(double beta) {
    // read lattice
    Lattice lattice(arg_latfile);
    alg = make_algorithm_objects(lattice);
    //double beta = 0.00001;
    bool arg_neg_init = false; //("z", "neg_init", "Negative initialization", false);

      for (size_t rep=0; rep < arg_nreps; rep++) {
        alg[rep].beta = beta;
        if (arg_neg_init) {
          alg[rep].set_negative();
        } else {
          alg[rep].reset_sites(rep);
        }
      }
    return 0;
}


int run(unsigned int arg_nsweeps, double dbeta) {

//    std::vector<std::pair<double, std::string> > en(arg_nreps);

#ifndef _OPENMP
    const unsigned n= 0;
#else
    #pragma omp parallel num_threads(n_threads)
    {
      const unsigned n = omp_get_thread_num();
#endif


      size_t r0 = arg_nreps * n / n_threads;
      size_t r1 = arg_nreps * (n + 1) / n_threads;


      //check to make sure the temperature update is legal.
      //set dbeta to 0 if illegal so T is not updated during
      //sweeps
      if (alg[0].beta + dbeta < 0.000001) {
	      dbeta = 0.0;
      }


      // main loop
      for (size_t rep = r0; rep < r1; rep++) {
        for (size_t sweep = 0; sweep < arg_nsweeps; ++sweep) {
          incr_current_beta(dbeta / arg_nsweeps, rep);
          alg[rep].do_sweep(sweep);
        }
//        alg[rep].get_energies(en, rep);
      }

#ifdef _OPENMP
    }
#endif

  return 0;

}


float get_acceptance_ratio() {
    int accepted_flips=0;
    int total_flips=0;
    for (size_t rep=0; rep<arg_nreps; rep++) {
	    accepted_flips += alg[rep].accepts;
	    total_flips += alg[rep].totals;
	    alg[rep].reset_acceptance();
    }
    
    return float(accepted_flips) / float(total_flips);
    
}


float get_average_energy() {
    float summ = 0;
    for (size_t rep=0; rep<arg_nreps; rep++) {
        summ += alg[rep].get_energy();
    }
    return summ / arg_nreps / alg[0].num_sites();
}


float get_average_absolute_magnetization() {
  float configSum=0;
  float this_config_sum;
  for (unsigned int rep=0; rep<arg_nreps; rep++) {
    //cout << rep << alg[rep].get_configuration();
    float spinSum=0;
    for (int i=0; i<alg[rep].num_sites(); i++){
      spinSum += alg[rep].get_site(i);
    }
    this_config_sum = abs(spinSum);
    configSum += this_config_sum / alg[rep].num_sites();
  }
  return configSum / alg.size();

}

int get_num_reps() {
  return arg_nreps;
}

int get_num_spins(){
  return alg[0].num_sites();
}


void get_all_energies(double* arr, int size) {
   for (int rep=0; rep<arg_nreps; rep++) {
       arr[rep] = alg[rep].get_energy();
   }


}

void print_lattice() {
   for (int rep=0; rep < arg_nreps; rep++) {
    for (int spin=0; spin<alg[rep].num_sites(); spin++) {
        cout << alg[rep].get_site(spin) << " ";
      }
      cout << endl; 
    }
  }
 
  

void get_lattice(double* arr, int size) {
  int i=0;
  for (int rep=0; rep<arg_nreps; rep++) {
    for (int spin=0; spin<alg[rep].num_sites(); spin++) {
      if (alg[rep].get_site(spin) == 1 ) {
        arr[i] = 1;
      } else {
        arr[i] = -1;
      }
      //arr[i] = alg[rep].get_site(spin);
     // if (spin==0) {arr[i] = 1;}
      i++;
    }
  }
}


void set_current_beta(double beta) {
  for (size_t rep=0; rep<arg_nreps; rep++) {
    alg[rep].beta = beta ;
  }
}

double clip_zero(double T) {
  if (T<0) {
    return 0;
  } else {
    return T;
  }
}


void incr_current_beta(double incr, size_t rep) {
  if (alg[rep].beta + incr > 0.000001) {
    alg[rep].beta += incr;
  }
}

double get_current_beta() {
  double _beta = 0;
  for (size_t rep=0; rep<arg_nreps; rep++) {
    _beta += alg[rep].beta;
  }
  _beta = _beta / arg_nreps;
  return _beta;
}
